lib/rust/ensogl/component/grid-view/src/lib.rs

```diff
@@ -0,0 +1,508 @@
+            let text_layer = text_layer.as_ref().and_then(|l| l.upgrade());
+            self.entry_creation_ctx.create_entry(&text_layer)
+        };
+        let entry = match visible_entries.entry((row, col)) {
+            Occupied(entry) => entry.into_mut(),
+            Vacant(lack_of_entry) => {
+                let new_entry = free_entries.pop().unwrap_or_else(create_new_entry);
+                set_entry_position(&new_entry, row, col, entry_size);
+                self.display_object.add_child(&new_entry);
+                lack_of_entry.insert(new_entry)
+            }
+        };
+        entry.frp().set_model(model);
+    }
+}
+
+
+
+// ================
+// === GridView ===
+// ================
+
+/// A template for [`GridView`] structure, where entry parameters and model are separate generic
+/// arguments.
+///
+/// It may be useful when using GridView in parametrized structs, where we want to avoid rewriting
+/// `Entry` bound in each place. Otherwise it's better to use [`GridView`].
+///
+/// Please note that some bounds are still required, as we use [`Widget`] and [`Frp`] nodes.
+#[derive(CloneRef, Debug, Deref, Derivative)]
+#[derivative(Clone(bound = ""))]
+pub struct GridViewTemplate<E: 'static, M: frp::node::Data, P: frp::node::Data> {
+    widget: Widget<Model<E, P>, Frp<M, P>>,
+}
+
+/// Grid View Component.
+///
+/// This Component displays any kind of entry `E` in a grid. To have it working, you need to
+/// * Set entries size ([`Frp::set_entries_size`]),
+/// * Declare (and keep up-to-date) the visible area ([`Frp::set_viewport`]),
+/// * Set up logic for providing models (see _Requesting for Models_ section).
+/// * Optionally entries parameters, if given entry does not have sensible default.
+/// * Finally reset the content, providing number of rows and columns ([`Frp::reset_entries`]).
+///
+/// # Positioning
+///
+/// Please mark, that this structure has its left-top corner docked to (0, 0) point of parent
+/// display object, as this is more intuitive way with handling grids.
+///
+/// # Entries Instantiation
+///
+/// The entry should implement [`Entry`] trait. Entries are instantiated lazily, only those visible
+/// in provided [`Frp::view_area`]. Once entries are no longer visible, are detached, but not
+/// dropped and may be re-used to display new entries when needed. This way we can achieve very
+/// efficient scrolling.
+///
+/// ## Requesting for Models
+///
+/// Once an entry is uncovered, the Grid View emits [`Frp::model_for_entry_needed`]. Then the proper
+/// model should be provided using [`Frp::model_for_entry`] endpoint - only then the entry will be
+/// displayed.
+///
+/// **Important**. The [`Frp::model_for_entry_needed`] are emitted once when needed and not repeated
+/// anymore, after adding connections to this FRP node in particular. Therefore make sure, that you
+/// connect providing models logic before emitting any of [`Frp::set_entries_size`] or
+/// [`Frp::set_viewport`].  
```

Comment:
Are you remembering all the data that was transmitted to you? If so, could you add a limit for it and remove old data? I'm just afraid that our memory consumption can blow up trough the window otherwise with big tables of data. What do you think? Maybethe worry is not needed now.

Reply:
No, the data are passed to visible entries. We don't cache them. When the view is scrolled back, we re-request them. I can add info there.